src/Node/Expression/TempNameExpression.php

```diff
@@ -13,19 +13,28 @@
 
 use Twig\Compiler;
 use Twig\Error\SyntaxError;
+use Twig\Node\Expression\Variable\LocalVariable;
 
 class TempNameExpression extends AbstractExpression
 {
+    static private int $salt = 0;
+
     public const RESERVED_NAMES = ['varargs', 'context', 'macros', 'blocks', 'this'];
 
-    public function __construct(string|int $name, int $lineno)
+    public function __construct(string|int|null $name, int $lineno)
     {
         // All names supported by ExpressionParser::parsePrimaryExpression() should be excluded
-        if (\in_array(strtolower($name), ['true', 'false', 'none', 'null'])) {
+        if ($name && \in_array(strtolower($name), ['true', 'false', 'none', 'null'])) {
             throw new SyntaxError(\sprintf('You cannot assign a value to "%s".', $name), $lineno);
         }
 
-        if (is_int($name) || ctype_digit($name)) {
+        if (self::class === static::class) {
+            trigger_deprecation('twig/twig', '3.15', 'The "%s" class is deprecated, use "%s" instead.', self::class, LocalVariable::class);
+        }
+
+        if (null === $name) {
+            $name = \sprintf('_l%d', self::$salt++);
```

Comment:
would such static state still allow to have reproducible template compilation ? I doubt it as compiling multiple templates in the same process will produce different local variable names.
The parser `varNameSalt` was introduced as a member variable precisely when working on such reproducible compilation, because that state is reset for each parsed template.

Reply:
I've moved the creation of the name at compilation time.